tools/patch.py

```diff
@@ -0,0 +1,208 @@
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+This is a basic patch tool that can parse unified diff format patches
+and apply them to files. It is not a full-featured patch tool and
+only supports a subset of the unified diff format, just enough to
+apply the patches we need. It won't properly validate the patch file
+format, so it's up to the user to ensure the patch file is correct.
+It's also not optimized for performance, so it may be slow for large
+patches or files.
+"""
+
+import os
+import sys
+from enum import Enum
+
+class FileAlreadyPatchedError(Exception):
+    pass
+
+class FileNotMatchingError(Exception):
+    pass
+
+class PatchMalformedError(Exception):
+    pass
+
+class LineType(Enum):
+    MATCH = ' '
+    ADD = '+'
+    REMOVE = '-'
+
+class PatchLine:
+    def __init__(self, line, line_type):
+        self.line = line
+        self.line_type = line_type
+
+    def __str__(self):
+        return f"{self.line_type.value} {self.line}"
+
+class Hunk:
+    def __init__(self, filename, location, count):
+        self.filename = filename
+        self.location = int(location)
+        self.count = int(count)
+        self.lines = []
+
+    def apply(self):
+        with open(self.filename, 'r') as file:
+            lines = file.readlines()
+
+        output = []
+        current_line = 1
+        relative_line = 0
+        for line in lines:
+            line = line.rstrip('\r\n')
+            # Pass through the lines not inside the hunk
+            if current_line < int(self.location) or current_line >= int(self.location + self.count):
```

Comment:
Perhaps my mental model is wrong, but would the `=>` mean the last hunk context line would be included?

Would it be more clear to make it strictly `>`? Either way, I think the outcome would be the same since the last hunk line will always be a `MATCH` type.

Reply:
So, if a hunk is of size 1 only, at location 100, and our current line is 100, then we're basically doing

```
if 100 < 100 or 100 > 101
```

to exclude the hunk information, which means we're going to be patching the one and only line 100 (count 1) from the hunk.

I can rewrite this logic to be a bit more readable, give me a minute.